src/exchanges/bitmex/BitmexMarketFeed.ts

```diff
@@ -0,0 +1,190 @@
+ * with the License. You may obtain a copy of the License at                                                               *
+ *                                                                                                                         *
+ * http://www.apache.org/licenses/LICENSE-2.0                                                                              *
+ *                                                                                                                         *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on     *
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the                      *
+ * License for the specific language governing permissions and limitations under the License.                              *
+ ***************************************************************************************************************************/
+import * as R from 'ramda';
+
+import { ExchangeFeed, ExchangeFeedConfig } from '../ExchangeFeed';
+import { SnapshotMessage, LevelMessage, TradeMessage } from '../../core/Messages';
+import { BITMEX_WS_FEED } from './BitmexCommon';
+import { Big } from '../../lib/types';
+import { OrderPool } from '../../lib/BookBuilder';
+import { Level3Order, PriceLevelFactory, PriceLevelWithOrders } from '../../lib/Orderbook';
+import {
+    BitmexMessage, OrderbookSnapshotMessage, OrderbookUpdateMessage, TradeMessage as BitmexTradeMessage,
+    TradeData, SubscriptionResponseMessage, PriceData, LevelUpdate,
+} from './BitmexInterfaces';
+
+export class BitmexMarketFeed extends ExchangeFeed {
+    readonly owner: string;
+    readonly feedUrl: string;
+    // Maps order IDs to the price that they exist at
+    private orderIdMap: { [orderId: number]: number };
+    // BitMEX WsAPI doesn't include a sequence number, so we have to keep track if it ourselves and hope for the best.
+    private seq: number;
+
+    constructor(config: ExchangeFeedConfig) {
+        if (!config.wsUrl) {
+            config.wsUrl = BITMEX_WS_FEED;
+        }
+        super(config);
+        this.owner = 'BitMEX';
+        this.feedUrl = config.wsUrl;
+        this.seq = 0;
+        this.connect();
+    }
+
+    public subscribe(productIds: string[]) {
+        this.logger.log('debug', `Subscribing to the following symbols: ${JSON.stringify(productIds)}`);
+        productIds.forEach((productId: string) => {
+            const subscribeMessage = {
+                op: 'subscribe',
+                args: [`orderBookL2:${productId}`, `trade:${productId}`],
+            };
+
+            this.send(JSON.stringify(subscribeMessage));
+        });
+    }
+
+    protected onOpen(): void {
+        // Nothing for now
+    }
+
+    protected handleMessage(rawMsg: string): void {
+        const msg: BitmexMessage = JSON.parse(rawMsg) as BitmexMessage;
+
+        if (msg.error) {
```

Comment:
The way we handle error messages in the feed (as opposed to network errors or runtime exceptions) is to pass an `ErrorMessage` instance down the feed. This allows downstream components to handle the message as they see fit.